File: packages/flutter_shared/lib/tools/text_stats_tool.dart
import 'dart:convert';
import 'text_stats_data.dart';

/// 文本统计工具
class TextStatsTool {
  TextStatsTool._();

  /// 统计结果
  static TextStatsData analyze(String input) {
    int charCount = input.length;
    int spaceCount = input.split('').where((c) => c == ' ' || c == '\t').length;
    int chineseCharCount = 0;
    int digitCount = 0;
    int letterCount = 0;
    int punctuationCount = 0;

    for (final char in input.runes) {
      if (char >= 0x4E00 && char <= 0x9FFF) {
        chineseCharCount++;
      } else if (char >= 0x3000 && char <= 0x303F) {
        chineseCharCount++;
      } else if (char >= 0xFF00 && char <= 0xFFEF) {
        chineseCharCount++;
      } else if (char >= 0x0030 && char <= 0x0039) {
        digitCount++;
      } else if ((char >= 0x0041 && char <= 0x005A) || (char >= 0x0061 && char <= 0x007A)) {
        letterCount++;
      } else if ((char >= 0x0021 && char <= 0x002F) ||
          (char >= 0x003A && char <= 0x0040) ||
          (char >= 0x005B && char <= 0x0060) ||
          (char >= 0x007B && char <= 0x007E)) {
        punctuationCount++;
      }
    }

    int charNoSpace = charCount - spaceCount;
    int wordCount = 0;
    if (input.trim().isNotEmpty) {
      wordCount = input.trim().split(RegExp(r'\s+')).length;
    }
    int lineCount = input.isEmpty ? 0 : '\n'.allMatches(input).length + 1;
    List<int> utf8Bytes = utf8.encode(input);
    int byteCountUtf8 = utf8Bytes.length;
    int byteCountUtf16 = input.length * 2;

    return TextStatsData(
      charCount: charCount,
      charNoSpace: charNoSpace,
      wordCount: wordCount,
      lineCount: lineCount,
      chineseCharCount: chineseCharCount,
      byteCountUtf8: byteCountUtf8,
      byteCountUtf16: byteCountUtf16,
      digitCount: digitCount,
      letterCount: letterCount,
      punctuationCount: punctuationCount,
      spaceCount: spaceCount,
    );
  }
}

File: miniapp/src/utils/tools/textStatsTool.ts
/**
 * 文本统计工具
 * 移植自 flutter_shared/tools/text_stats_tool.dart
 */

export interface TextStatsData {
  charCount: number;
  charNoSpace: number;
  wordCount: number;
  lineCount: number;
  chineseCharCount: number;
  byteCountUtf8: number;
  byteCountUtf16: number;
  digitCount: number;
  letterCount: number;
  punctuationCount: number;
  spaceCount: number;
}

export const TextStatsTool = {
  analyze(input: string): TextStatsData {
    let charCount = input.length;
    let spaceCount = 0;
    let chineseCharCount = 0;
    let digitCount = 0;
    let letterCount = 0;
    let punctuationCount = 0;

    for (const ch of input) {
      const code = ch.charCodeAt(0);
      if (ch === ' ' || ch === '\t') {
        spaceCount++;
      } else if ((code >= 0x4E00 && code <= 0x9FFF) ||
                 (code >= 0x3000 && code <= 0x303F) ||
                 (code >= 0xFF00 && code <= 0xFFEF)) {
        chineseCharCount++;
      } else if (code >= 0x0030 && code <= 0x0039) {
        digitCount++;
      } else if ((code >= 0x0041 && code <= 0x005A) || (code >= 0x0061 && code <= 0x007A)) {
        letterCount++;
      } else if ((code >= 0x0021 && code <= 0x002F) ||
                 (code >= 0x003A && code <= 0x0040) ||
                 (code >= 0x005B && code <= 0x0060) ||
                 (code >= 0x007B && code <= 0x007E)) {
        punctuationCount++;
      }
    }

    const charNoSpace = charCount - spaceCount;
    const wordCount = input.trim() ? input.trim().split(/\s+/).length : 0;
    const lineCount = input ? input.split('\n').length : 0;
    const encoder = new TextEncoder();
    const byteCountUtf8 = encoder.encode(input).length;
    const byteCountUtf16 = input.length * 2;

    return {
      charCount,
      charNoSpace,
      wordCount,
      lineCount,
      chineseCharCount,
      byteCountUtf8,
      byteCountUtf16,
      digitCount,
      letterCount,
      punctuationCount,
      spaceCount,
    };
  },
};
